### PR TITLE
C2: Call to compute_separating_interferences has wrong argument order

### PR DESCRIPTION
**Pull request by r-v-raghav**
**# [.../src/hotspot/share/opto/coalesce.hpp] -**
**112** uint compute_separating_interferences(Node *dst_copy, Node *src_copy, Block *b, uint bindex, RegMask &rm, **uint rm_size, uint reg_degree,** uint lr1, uint lr2);

**# [.../src/hotspot/share/opto/coalesce.cpp] -**
..........
**538** uint PhaseConservativeCoalesce::compute_separating_interferences(Node *dst_copy, Node *src_copy, Block *b, uint bindex, RegMask &rm, **uint reg_degree, uint rm_size,** uint lr1, uint lr2 ) {
.................
**747** reg_degree = compute_separating_interferences(dst_copy, src_copy, b, bindex, rm, **rm_size, reg_degree,** lr1, lr2 );

**# So fixing the argument order - [.../src/hotspot/share/opto/coalesce.cpp]**
-uint PhaseConservativeCoalesce::compute_separating_interferences(Node *dst_copy, Node *src_copy, Block *b, uint bindex, RegMask &rm, **uint reg_degree, uint rm_size,** uint lr1, uint lr2 ) {
+uint PhaseConservativeCoalesce::compute_separating_interferences(Node *dst_copy, Node *src_copy, Block *b, uint bindex, RegMask &rm, **uint rm_size, uint reg_degree,** uint lr1, uint lr2 ) {

### CODE REVIEW

**Comment by shipilev** [comment]
This change looks good, but my question is more wide: does it actually break/fixes any compiler bug? I would have thought changing RA would have some impact. In other words, are there test failures on higher tiers?

**Review by TobiHartmann — APPROVED** [review]
Looks good to me but I'm also wondering why that was never an issue and if the fix could now trigger previously hidden problems. We should at least give it a good amount of testing before integration.

**Review by vnkozlov — APPROVED** [review]

**Comment by vnkozlov** [comment]
> Looks good to me but I'm also wondering why that was never an issue and if the fix could now trigger previously hidden problems. We should at least give it a good amount of testing before integration.

compute_separating_interferences() is used only in one place and resulting `reg_degree` is only used in place to compare with `max_juint`:
https://github.com/openjdk/jdk/blob/master/src/hotspot/share/opto/coalesce.cpp#L746
Due to this bug the returning value is most likely `max_juint` and we bailout coalesce of copies. As result we would have more LRG live ranges to work with but Graph still can be processed.
In short RA produce less optimal code.
I agree with @TobiHartmann we should do a lot of tiers testing for this fix because coalescing code after L755 may be not exercised as much as we want due to this bug.

**Comment by r-v-raghav** [comment]
Thanks for the details @shipilev, @TobiHartmann, @vnkozlov.
I too could not link any other unresolved or 'won't-fix/incomplete' old c2-regalloc bug to this case.
Yes I will initiate more tier testing with the change before integeration.
